uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChangeTest.php

```diff
@@ -1786,5 +1899,257 @@ public function testGetTotalTOILBalanceChangeForContactWithinAGivenPeriodAndWith
   }
+
+  public function testRecalculateExpiredBalanceChangesForLeaveRequestPastDates() {
+    $absencePeriod = AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('-50 days'),
+      'end_date' => CRM_Utils_Date::processDate('-20 days')
+    ]);
+
+    $periodEntitlement1 = LeavePeriodEntitlementFabricator::fabricate([
+      'contact_id' => 1,
+      'period_id' => $absencePeriod->id,
+      'type_id' => 1,
+    ]);
+
+    $periodEntitlement2 = LeavePeriodEntitlementFabricator::fabricate([
+      'contact_id' => 1,
+      'period_id' => $absencePeriod->id,
+      'type_id' => 2,
+    ]);
+
+    $leaveRequestStatuses = array_flip(LeaveRequest::buildOptions('status_id'));
+
+    $balanceChange1 = $this->createExpiredBroughtForwardBalanceChange(
+      $periodEntitlement1->id,
+      1,
+      1,
+      30
+    );
+
+    //This balance change expired before the leave request dates so it will not be affected
+    $balanceChange2 = $this->createExpiredBroughtForwardBalanceChange(
+      $periodEntitlement2->id,
+      5,
+      5,
+      32
+    );
+
+    $balanceChange3 = $this->createExpiredTOILRequestBalanceChange(
+      $periodEntitlement1->type_id,
+      $periodEntitlement1->contact_id,
+      $leaveRequestStatuses['Approved'],
+      CRM_Utils_Date::processDate('-40 days'),
+      CRM_Utils_Date::processDate('-39 days'),
+      5,
+      CRM_Utils_Date::processDate('-29 days'),
+      5
+    );
+
+    // A leave request with past dates with two of the dates before
+    // an expired brought forward and an expired TOIL
+    $leaveRequest = $this->createLeaveRequestBalanceChange(
+      $periodEntitlement1->type_id,
+      $periodEntitlement1->contact_id,
+      $leaveRequestStatuses['Approved'],
+      CRM_Utils_Date::processDate('-31 days'),
+      CRM_Utils_Date::processDate('-29 days')
+    );
+
+    $numberOfUpdatedRecords = LeaveBalanceChange::recalculateExpiredBalanceChangesForLeaveRequestPastDates($leaveRequest);
+    $this->assertEquals(2, $numberOfUpdatedRecords);
+
+    $expiryRecord1 = new LeaveBalanceChange();
+    $expiryRecord1->id = $balanceChange1->id;
+    $expiryRecord1->find(true);
+
+    $expiryRecord2 = new LeaveBalanceChange();
+    $expiryRecord2->id = $balanceChange2->id;
+    $expiryRecord2->find(true);
+
+    $expiryRecord3 = new LeaveBalanceChange();
+    $expiryRecord3->id = $balanceChange3->id;
+    $expiryRecord3->find(true);
+
+    //The first day of the leave request is 31 days ago and will recalculate balanceChange1 (brought forward expired)
+    //since it expired first before the TOIL. It will deduct one day remaining 0
+    //The second day of leave request is 30 days ago and it will deduct one from expired TOIL balance change remaining 4
+    //The third day of leave request is 29 days ago and will deduct one also since the TOIL expired on that very day remaining 3
+    //expiredbalanceChange2 is untouched because it expired before the leave request dates
```

Comment:
Even though this is true, it's not the main reason for it to be left untouched. There's also the point that balance change 2 is linked to another leave period entitlement and absence type. So maybe we should also add leave requests to that absence type and check that it recalculates things right based on it?